sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -1207,6 +1207,17 @@ class Analyzer(
                 val alias = Alias(ae, ae.toString)()
                 aggregateExpressions += alias
                 alias.toAttribute
+              // Replacing [[NamedExpression]] causes the error on [[Grouping]] because the
+              // grouping column will be new attribute created by adding additional [[Alias]].
+              // So we can't find the grouping column and replace it in the rule
+              // [[ResolveGroupingAnalytics]].
```

Comment:
When there is a `Grouping(col#1)` in the HAVING clause. If we push down it to Aggregate and use an Alias instead, the grouping function become `Grouping(col#1#2)`. Then in the rule `ResolveGroupingAnalytics` we try to find the index of col in group by expressions, we can't find it.